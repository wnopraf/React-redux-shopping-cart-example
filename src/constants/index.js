 export const ADD_PRODUCT = 'ADD_PRODUCT', DELETE_PRODUCT = 'DELETE_PRODUCT', INCREMENT_STOCK = 'INCREMENT_SOTCK', DECREMENT_SOTCK = 'DECREMENT_STOCK', INCREMENT_AMOUNT = 'INCREMENT_AMOUNT', DECREMENT_AMOUNT = 'DRECREMENT_AMOUNT'
 