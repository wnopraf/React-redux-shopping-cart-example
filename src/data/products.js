export default {
  'products': [
    {
      'id': 1,
      'name': 'product-A',
      'prize': '68.00'
    },
    {
      'id': 2,
      'name': 'product-B',
      'prize': '114.10'
    },
    {
      'id': 3,
      'name': 'product-C',
      'prize': '66.02'
    },
    {
      'id': 4,
      'name': 'product-D',
      'prize': '135.66'
    },
    {
      'id': 5,
      'name': 'product-E',
      'prize': '136.68'
    },
    {
      'id': 6,
      'name': 'product-F',
      'prize': '102.56'
    },
    {
      'id': 7,
      'name': 'product-G',
      'prize': '112.90'
    },
    {
      'id': 8,
      'name': 'product-H',
      'prize': '92.23'
    },
    {
      'id': 9,
      'name': 'product-I',
      'prize': '73.13'
    },
    {
      'id': 10,
      'name': 'product-J',
      'prize': '137.39'
    },
    {
      'id': 11,
      'name': 'product-K',
      'prize': '129.77'
    },
    {
      'id': 12,
      'name': 'product-E',
      'prize': '81.40'
    },
    {
      'id': 13,
      'name': 'product-L',
      'prize': '107.21'
    },
    {
      'id': 14,
      'name': 'product-M',
      'prize': '147.86'
    },
    {
      'id': 15,
      'name': 'product-N',
      'prize': '111.09'
    },
    {
      'id': 16,
      'name': 'product-Ã‘',
      'prize': '133.72'
    },
    {
      'id': 17,
      'name': 'product-O',
      'prize': '141.38'
    },
    {
      'id': 18,
      'name': 'product-P',
      'prize': '103.80'
    },
    {
      'id': 19,
      'name': 'product-Q',
      'prize': '74.59'
    },
    {
      'id': 20,
      'name': 'product-R',
      'prize': '69.00'
    },
    {
      'id': 21,
      'name': 'product-S',
      'prize': '111.35'
    },
    {
      'id': 22,
      'name': 'product-T',
      'prize': '107.80'
    },
    {
      'id': 23,
      'name': 'product-U',
      'prize': '135.51'
    },
    {
      'id': 24,
      'name': 'product-V',
      'prize': '101.85'
    },
    {
      'id': 25,
      'name': 'product-W',
      'prize': '67.34'
    },
    {
      'id': 26,
      'name': 'product-Y',
      'prize': '120.43'
    },
    {
      'id': 27,
      'name': 'product-Z',
      'prize': '69.00'
    }
  ],
  'stock': [
    {
      'productId': 1,
      'stock': 13
    },
    {
      'productId': 2,
      'stock': 10
    },
    {
      'productId': 3,
      'stock': 10
    },
    {
      'productId': 4,
      'stock': 13
    },
    {
      'productId': 5,
      'stock': 13
    },
    {
      'productId': 6,
      'stock': 13
    },
    {
      'productId': 7,
      'stock': 8
    },
    {
      'productId': 8,
      'stock': 13
    },
    {
      'productId': 9,
      'stock': 13
    },
    {
      'productId': 10,
      'stock': 9
    },
    {
      'productId': 11,
      'stock': 10
    },
    {
      'productId': 12,
      'stock': 9
    },
    {
      'productId': 13,
      'stock': 10
    },
    {
      'productId': 14,
      'stock': 10
    },
    {
      'productId': 15,
      'stock': 9
    },
    {
      'productId': 16,
      'stock': 13
    },
    {
      'productId': 17,
      'stock': 12
    },
    {
      'productId': 18,
      'stock': 14
    },
    {
      'productId': 19,
      'stock': 12
    },
    {
      'productId': 20,
      'stock': 11
    },
    {
      'productId': 21,
      'stock': 15
    },
    {
      'productId': 22,
      'stock': 14
    },
    {
      'productId': 23,
      'stock': 9
    },
    {
      'productId': 24,
      'stock': 8
    },
    {
      'productId': 25,
      'stock': 14
    },
    {
      'productId': 26,
      'stock': 9
    },
    {
      'productId': 27,
      'stock': 15
    }
  ]
}
