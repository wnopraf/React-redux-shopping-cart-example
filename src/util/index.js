export const selectIdItem = (stateSlice, id) => stateSlice.find(e => e.productId === id)
