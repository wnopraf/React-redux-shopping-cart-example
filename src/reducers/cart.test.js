import { INCREMENT_AMOUNT, DECREMENT_AMOUNT } from '../constants'
import { selectIdItem } from '../util'
import amountReducer from './cart'
import data from '../data/products'
const { products } = data

const mockCart = products.slice(2, 9).map(({ id }) => ({ productId: id, amount: Math.floor(Math.random() * 6 + 1) }))
describe('amount reducer', () => {
  const incAction = {
    type: INCREMENT_AMOUNT,
    id: 7
  }
  const decAction = {
    type: DECREMENT_AMOUNT,
    id: 7
  }
  describe('Empty cart', () => {
    test('Initial cart item', () => {
      const initAmount = amountReducer(undefined, incAction)[0]
      expect(initAmount).toEqual({ productId: 7, amount: 1 })
    })
    test('Increment action', () => {
      const amountState = [{ productId: 7, amount: 4 }]
      const inAmount = amountReducer(amountState, incAction)
      expect(inAmount).toEqual([{ productId: 7, amount: 5 }])
    })
    test('Increment twice', () => {
      const amountState = [{ productId: 7, amount: 4 }]
      const inAmount = amountReducer(amountState, incAction)
      expect(inAmount).not.toEqual([{ productId: 7, amount: 6 }])
    })
    test('Decrement Action', () => {
      const amountState = [{ productId: 7, amount: 8 }]
      const decAmount = amountReducer(amountState, decAction)
      expect(decAmount).toEqual([{ productId: 7, amount: 7 }])
    })
  })
  describe('Mock cart with data', () => {
    test('Mock cart length', () => {
      expect(mockCart).toHaveLength(7)
    })
    describe('Incrementing item amount', () => {
      const incAction = {
        type: INCREMENT_AMOUNT,
        id: 3
      }
      it.skip('Cart state', () => {
        expect(mockCart).toMatchSnapshot()
      })
      const itemBeforeAction = selectIdItem(mockCart, 3)
      const cartAfterAction = amountReducer(mockCart, incAction)
      const itemAfterAction = selectIdItem(cartAfterAction, 3)
      it('Has the right length after increment', () => {
        expect(cartAfterAction).toHaveLength(7)
      })
      it('Has the right amount after incrment', () => {
        expect(itemAfterAction.amount).toBe(itemBeforeAction.amount + 1)
      })
    })
    describe('adding new item to cart', () => {
      const incAction = {
        type: INCREMENT_AMOUNT,
        id: 12
      }
      const reducerState = amountReducer(mockCart, incAction)
      it('Cart has has the right length', () => {
        expect(reducerState).toHaveLength(8)
      })
      it('New Cart item has the rigth amount', () => {
        const newItem = selectIdItem(reducerState, incAction.id)
        expect(newItem.amount).toBe(1)
      })
    })
    describe('Deleting item from cart', () => {
      const delAction = {
        type: DECREMENT_AMOUNT,
        id: 12
      }
      const inAction = {
        type: INCREMENT_AMOUNT,
        id: 12
      }
      let cart = amountReducer(mockCart, inAction)
      it('Cart has 8 items before deleting', () => {
        expect(cart).toHaveLength(8)
      })
      it('Item has amount 1 before deleting', () => {
        const { amount } = selectIdItem(cart, 12)
        expect(amount).toBe(1)
      })
      it('Cart has 7 items after deleting', () => {
        cart = amountReducer(cart, delAction)
        expect(cart).toHaveLength(7)
      })
    })
  })
})
