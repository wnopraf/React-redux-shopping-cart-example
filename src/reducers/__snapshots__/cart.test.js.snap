// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amount reducer Another item to cart Cart state 1`] = `
Array [
  Object {
    "amount": 6,
    "productId": 3,
  },
  Object {
    "amount": 1,
    "productId": 4,
  },
  Object {
    "amount": 3,
    "productId": 5,
  },
  Object {
    "amount": 4,
    "productId": 6,
  },
  Object {
    "amount": 2,
    "productId": 7,
  },
  Object {
    "amount": 1,
    "productId": 8,
  },
  Object {
    "amount": 3,
    "productId": 9,
  },
]
`;
